aspnetcore/fundamentals/servers/kestrel.md

```diff
@@ -1019,34 +1344,157 @@ TLS restrictions for HTTP/2:
 The following example permits HTTP/1.1 and HTTP/2 connections on port 8000. Connections are secured by TLS with a supplied certificate:
 
 ```csharp
-public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
-    WebHost.CreateDefaultBuilder(args)
-        .UseStartup<Startup>()
-        .ConfigureKestrel((context, options) =>
-        {
-            options.Listen(IPAddress.Any, 8000, listenOptions =>
-            {
-                listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
-                listenOptions.UseHttps("testCert.pfx", "testPassword");
-            });
-        });
+.ConfigureKestrel(serverOptions =>
+{
+    serverOptions.Listen(IPAddress.Any, 8000, listenOptions =>
+    {
+        listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
```

Comment:
I might be missing something here, but these remarks seem contradictory where "non" is used. Is "non-HTTPS" the same as "HTTP"? If so, then these remarks don't seem to make sense.

Reply:
I'm using non-HTTPS to mean non-encrypted regardless of HTTP protocol version. A URL with `https://` could either reference an HTTP/1.x or HTTP/2 endpoint (or one that supports both via ALPN as Kestrel does by default). Non-HTTPS means that communication with the endpoint is *not* encrypted with TLS.

In ASP.NET Core 3.0, Kestrel will default to supporting both HTTP/1.x and HTTP/2 for encrypted endpoints. This means Kestrel effectively defaults to HTTP/2 for clients that support it. This is different from saying "The default protocol is HTTP/1.1."

HTTP/2 is not the default for endpoints that are not encrypted because browsers always attempt to make HTTP/1.1 requests to any non-encrypted endpoint as there's no ALPN to upgrade to HTTP/2 without prior knowledge. You can however support HTTP/2 on a non-encrypted endpoint if you opt in to it because you know clients will have prior knowledge that the endpoint supports HTTP/2.

It's not possible to support both HTTP/1.x and HTTP/2 on a non-encrypted endpoint because the lack of ALPN. Does that clear things up?

Reply:
@halter73 *Naaaaaah* ... nevermind those earlier rubber :duck: remarks.

Take a look at the `ListenOptions.Protocols` section. Doesn't that do a fine job in its present form?

[Direct link to ListenOptions.Protocols section](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-2.1&branch=pr-en-us-13775#listenoptionsprotocols)